src/Stratis.Bitcoin.Features.BlockStore/BlockStore.cs

```diff
@@ -0,0 +1,363 @@
+                dequeueTask = dequeueTask ?? this.blocksQueue.DequeueAsync();
+
+                // Wait for one of the tasks: dequeue or timer (if available) to finish.
+                Task task = (timerTask == null) ? dequeueTask : await Task.WhenAny(dequeueTask, timerTask).ConfigureAwait(false);
+
+                bool saveBatch = false;
+
+                try
+                {
+                    await task.ConfigureAwait(false);
+                }
+                catch (OperationCanceledException)
+                {
+                    // Happens when node is shutting down or Dispose() is called.
+                    // We want to save whatever is in the batch before exiting the loop.
+                    saveBatch = true;
+
+                    this.logger.LogDebug("Node is shutting down. Save batch.");
+                }
+
+                // Save batch if timer ran out or we've dequeued a new block and reached the consensus tip 
+                // or the max batch size is reached or the node is shutting down.  
+                if (dequeueTask.Status == TaskStatus.RanToCompletion)
+                {
+                    BlockPair item = dequeueTask.Result;
+
+                    // Set the dequeue task to null so it can be assigned on the next iteration.
+                    dequeueTask = null;
+
+                    batch.Add(item);
+
+                    this.currentBatchSizeBytes += item.Block.GetSerializedSize();
+
+                    saveBatch |= (item.ChainedHeader == this.chain.Tip) || (this.currentBatchSizeBytes >= BatchThresholdSizeBytes);
+                }
+                else
+                {
+                    // Will be executed in case timer ran out or node is being shut down.
+                    saveBatch = true;
+                }
+
+                if (saveBatch)
+                {
+                    if (batch.Count != 0)
+                    {
+                        await this.SaveBatchAsync(batch).ConfigureAwait(false);
+
+                        batch.Clear();
+                        this.currentBatchSizeBytes = 0;
+                    }
+
+                    timerTask = null;
+                }
+                else
+                {
+                    // Start timer if it is not started already.
+                    timerTask = timerTask ?? Task.Delay(BatchMaxSaveIntervalSeconds * 1000, this.nodeLifetime.ApplicationStopping);
+                }
+            }
+
+            this.logger.LogTrace("(-)");
+        }
+
+        /// <summary>Checks if repository contains reorged blocks and deletes them; saves batch on top. </summary>
+        /// <param name="batch">List of batched blocks. Cannot be empty.</param>
+        private async Task SaveBatchAsync(List<BlockPair> batch)
+        {
+            this.logger.LogTrace("({0}.{1}:{2})", nameof(batch), nameof(batch.Count), batch.Count);
+
+            List<BlockPair> batchCleared = this.GetBatchWithoutReorgedBlocks(batch);
```

Comment:
I thought i mentioned this but maybe I missed it.... batchCleared sounds wrong, is it meant to be clearedBatch?

Reply:
`clearedBatch` is just a short name because it's handy. There is no need to have a long name since code misunderstanding is avoided by having `GetBatchWithoutReorgedBlocks` for a method name + it being well commented.

Reply:
Yeah ok so clearedBatch rather than batchCleared then

Reply:
I think this one is about the personal preference. For me `batchCleared` sounds better. 
If there is anyone else who would like to have this change- let me know, I don't mind changing that in case majority is for it.

Reply:
clearedBatch is like a noun of a batch that was cleared, batchCleared sounds like the event that happened . I don't have a preference over word order if it means the same, but feels different to me here.

Reply:
As Andy is native speaker, I wouldn't really argue with him what sounds better there and just go with whatever he suggests here. It is kind of irrelevant what sounds better to us. Our English is not on that level.

Reply:
That might sound weird, so I rather explain what I meant. What I meant was that if the name selection was purely done on what sounds better then we should go with native speakers recommendation. If there is some kind of logic behind that, however, it should be explained and defended.

Reply:
Good point. Fixed